packages/service-library/src/servicelib/application_setup.py

```diff
@@ -0,0 +1,154 @@
+    :param config_enabled: option in config to enable, defaults to None which is '$(module-section).enabled' (config_section and config_enabled are mutually exclusive)
+    :raises DependencyError
+    :raises AppSetupBaseError
+    :return: False if setup was skipped
+    :rtype: bool
+
+    :Example:
+        from servicelib.application_setup import app_module_setup
+
+        @app_module_setup('mysubsystem', ModuleCategory.SYSTEM, logger=log)
+        def setup(app: web.Application):
+            ...
+    """
+    # TODO: resilience to failure. if this setup fails, then considering dependencies, is it fatal or app can start?
+
+    module_name = module_name.replace(".__init__", "")
+    depends = depends or []
+
+    if config_section and config_enabled:
+        raise ValueError("Can only set config_section or config_enabled but not both")
+
+    section = config_section or module_name.split(".")[-1]
+    if config_enabled is None:
+        config_enabled = f"{section}.enabled"
+    else:
+        # if passes config_enabled, invalidates info on section
+        section = None
+
+    logger = logger or log
+
+    def decorate(setup_func):
+
+        if "setup" not in setup_func.__name__:
+            logger.warning("Rename '%s' to contain 'setup'", setup_func.__name__)
+
+       # metadata info
+        def setup_metadata() -> Dict:
+            return {
+                'module_name': module_name,
+                'dependencies': depends,
+                'config_section': section,
+                'config_enabled': config_enabled
+            }
+
+        # wrapper
+        @functools.wraps(setup_func)
+        def setup_wrapper(app: web.Application, *args, **kargs) -> bool:
+            # pre-setup
+            logger.debug("Setting up '%s' [%s; %s] ... ", module_name, category.name, depends)
+
+            if APP_SETUP_KEY not in app:
+                app[APP_SETUP_KEY] = []
+
+            if category == ModuleCategory.ADDON:
+                # NOTE: ONLY addons can be enabled/disabled
+                # TODO: sometimes section is optional, check in config schema
+                cfg = app[APP_CONFIG_KEY]
+
+                def _get(cfg_, parts):
+                    for part in parts:
+                        if section and part == "enabled": # if section exists, no need to explicitly enable it
+                            cfg_ = cfg_.get(part, True)
+                        else:
+                            cfg_ = cfg_[part]
+                    return cfg_
+
+                try:
+                    is_enabled = _get(cfg, config_enabled.split("."))
+                except KeyError as ee:
+                    raise AppSetupBaseError(f"Cannot find '{config_enabled}' in app config in [ {ee} ]")
+
+                if not is_enabled:
+                    logger.info("Skipping '%s' setup. Explicitly disabled in config", module_name)
+                    return False
```

Comment:
how do you separate skipped setup from error? or is False always a skipped setup?

Reply:
Good point. *For the moment* an exception is an error, False means skipped and True means setup ran w/o problems. 
Since there are multiple states, the right way to go might be using exceptions but this solution is intented to keep legacy in all tests until #1142 is complete